### DISCUSSION TITLE
Return text fields into which the text cannot be inserted completely

### DISCUSSION THREAD

**Discussion by thomasbaustert** [gettalong/hexapdf#289] [Ideas]
If a text field of a PDF template is filled with a text, it can happen that the text does not fit completely into the field.
It would be very helpful to be informed in which fields the text has not been inserted completely.

For now I tried the following:

```ruby
def text_fits?(field)
  x_bottom_left, y_bottom_left, x_top_right, y_top_right = field[:Rect]
  field_height = y_top_right - y_bottom_left
  field_width = x_top_right - x_bottom_left

  _, font_size, = field.parse_default_appearance_string

  # prawn_document = Prawn::Document.new(...)
  text_width = prawn_document.width_of(field.field_value, size: font_size)

  (text_width - field_width < 0)
end
```

I use Prawn's method `width_of` to get the width of the text (based on the font and the font size).
If the text width is less then the text field width the texts fits.

When I look at the generated PDF, I see that if `text_fits?` returns true, the text is inserted completely. And if `text_fits?` returns false, the text cannot be inserted completely.

Do you think this is reliable way?
Is there a similar method like `width_of` in hexapdf?
Or is there a better way in hexapdf to achieve this?

For example, something similar to Prawn's `formatted_text_box` which returns the text that does not fit into the box:

```ruby
# remaining = prawn_doc.formatted_text_box(text, ...)
remaining = (field.field_value = text)
if remaining.empty?
  puts "fits"
else
  puts "fits not, remaining=`#{remaining}`"
end
```

Thanks!

**Comment by gettalong** [comment]
There is currently no way to determine this in HexaPDF.

Due to the way your `text_fits?` method is implemented it will not work for all cases because e.g. the rectangle may be rotated, there may be a border, and Prawn's `width_of` method is not fully compatible with the way HexaPDF lays out text.

I will have a look at how to best provide that information.

**Comment by thomasbaustert** [comment]
Thanks for having a look into it! 
The PDF templates are generated by us (by a client of mine to be precisely). 
For now there is no rotation or border, just a simple text field. Maybe it makes a solution easier.
